File: Frontend/greenflora/types/market.ts
/**
 * types/market.ts
 *
 * TypeScript shapes for the Market Intelligence feature.
 * Mirrors the backend schemas in Backend/schemas/market.py.
 * All values originate from the AMIS-ingested Supabase data —
 * fields are null/empty when the underlying data is missing.
 */

/** One selectable crop with its latest known price snapshot. */
export interface MarketCommodity {
  id: string;
  name: string;
  category: string | null;
  unit: string | null;
  /** Most recent date (ISO) this crop had any reported price. */
  latest_date: string | null;
  /** Representative price (FQP-based) on latest_date. */
  latest_price: number | null;
  /** How many markets reported a price on latest_date. */
  markets_reporting: number;
}

/** Response of GET /api/market/commodities. */
export interface MarketCommoditiesResponse {
  commodities: MarketCommodity[];
  total: number;
  /** False when the AMIS pipeline has not ingested any data yet. */
  data_available: boolean;
}

/** Price of one crop in one market on the latest date. */
export interface MarketComparisonEntry {
  market_id: string;
  name: string;
  price: number;
  min_price: number | null;
  max_price: number | null;
  quantity: number | null;
  date: string | null;
}

/** One point of the daily price trend series. */
export interface MarketTrendPoint {
  date: string;
  price: number;
}

/** Share of total arrivals (quantity) for one market. */
export interface MarketDistributionEntry {
  market_id: string;
  name: string;
  quantity: number;
  share_pct: number;
}

/** Arrivals distribution across markets for the latest date. */
export interface MarketDistribution {
  entries: MarketDistributionEntry[];
  total_quantity: number;
}

/** Simple data-driven price signal. */
export type MarketSignal =
  | "rising"
  | "falling"
  | "stable"
  | "insufficient_data";

/** How the representative current price was derived. */
export type PriceBasis = "weighted" | "average" | "single" | "unknown";

/** Response of GET /api/market/overview. */
export interface MarketOverview {
  commodity_id: string;
  commodity_name: string;
  category: string | null;
  /** Price unit, e.g. "Rs/100Kg". */
  unit: string | null;

  // Data coverage
  latest_date: string | null;
  first_date: string | null;
  /** Number of distinct dates with price data in the fetched window. */
  days_of_data: number;
  markets_reporting: number;

  // Summary cards
  current_price: number | null;
  price_basis: PriceBasis;
  change_pct: number | null;
  change_period_days: number | null;
  signal: MarketSignal;
  highest_market: MarketComparisonEntry | null;
  lowest_market: MarketComparisonEntry | null;
  spread_abs: number | null;
  spread_pct: number | null;

  // Charts
  trend: MarketTrendPoint[];
  /** Market the trend is scoped to; null means all-market average. */
  trend_market_id: string | null;
  market_comparison: MarketComparisonEntry[];
  distribution: MarketDistribution | null;

  // Farmer insights
  insights: string[];
}

/** Available time-period filters for the trend chart. */
export type MarketPeriod = "7D" | "30D" | "3M" | "6M";

/** Number of days each period filter spans (matches backend window). */
export const PERIOD_DAYS: Record<MarketPeriod, number> = {
  "7D": 7,
  "30D": 30,
  "3M": 90,
  "6M": 180,
};

File: Backend/schemas/market.py
"""
market.py (schemas)

API-facing response schemas for the Market Intelligence endpoints.

All values come from the AMIS-ingested Supabase tables
(``commodities``, ``markets``, ``crop_market_rates``) — never fabricated.
Fields are ``None`` / empty when the underlying data is missing so the
frontend can render honest empty states.
"""

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Commodities list (crop selector)
# ---------------------------------------------------------------------------

class MarketCommodityItem(BaseModel):
    """One selectable crop with its latest known price snapshot."""

    id: str
    name: str
    category: Optional[str] = None
    unit: Optional[str] = None
    # Most recent date this crop had any reported price (ISO date).
    latest_date: Optional[str] = None
    # Representative price (FQP-based) on latest_date.
    latest_price: Optional[float] = None
    # How many markets reported a price on latest_date.
    markets_reporting: int = 0


class MarketCommoditiesResponse(BaseModel):
    """Response for the crop-selector list endpoint."""

    commodities: list[MarketCommodityItem] = []
    total: int = 0
    # False when the AMIS pipeline has not ingested any data yet.
    data_available: bool = True


# ---------------------------------------------------------------------------
# Market overview (single commodity)
# ---------------------------------------------------------------------------

class MarketComparisonEntry(BaseModel):
    """Price of one crop in one market on the latest date."""

    market_id: str
    name: str
    price: float
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    quantity: Optional[float] = None
    date: Optional[str] = None


class MarketTrendPoint(BaseModel):
    """One point of the price trend series (daily)."""

    date: str
    price: float


class MarketDistributionEntry(BaseModel):
    """Share of total arrivals (quantity) for one market."""

    market_id: str
    name: str
    quantity: float
    share_pct: float


class MarketDistribution(BaseModel):
    """Arrivals distribution across markets for the latest date."""

    entries: list[MarketDistributionEntry] = []
    total_quantity: float = 0.0


class MarketSignal(str, Enum):
    RISING = "rising"
    FALLING = "falling"
    STABLE = "stable"
    INSUFFICIENT_DATA = "insufficient_data"


class MarketOverviewResponse(BaseModel):
    """Complete market-intelligence bundle for one crop."""

    commodity_id: str
    commodity_name: str
    category: Optional[str] = None
    # Price unit, e.g. "Rs/100Kg".
    unit: Optional[str] = None

    # -- Data coverage -------------------------------------------------------
    latest_date: Optional[str] = None
    first_date: Optional[str] = None
    # Number of distinct dates with price data in the fetched window.
    days_of_data: int = 0
    markets_reporting: int = 0

    # -- Summary cards -------------------------------------------------------
    current_price: Optional[float] = None
    # How the current price was derived (shown to keep the UI honest).
    price_basis: str = "unknown"
    change_pct: Optional[float] = None
    change_period_days: Optional[int] = None
    signal: MarketSignal = MarketSignal.INSUFFICIENT_DATA
    highest_market: Optional[MarketComparisonEntry] = None
    lowest_market: Optional[MarketComparisonEntry] = None
    spread_abs: Optional[float] = None
    spread_pct: Optional[float] = None

    # -- Charts --------------------------------------------------------------
    trend: list[MarketTrendPoint] = []
    # Market the trend is scoped to; None means all-market average.
    trend_market_id: Optional[str] = None
    market_comparison: list[MarketComparisonEntry] = []
    distribution: Optional[MarketDistribution] = None

    # -- Farmer insights -----------------------------------------------------
    insights: list[str] = Field(
        default_factory=list,
        description="Short, data-driven, farmer-friendly insight strings.",
    )
